### PR TITLE
Admin

### PR DESCRIPTION
**Pull request by tuannguyen1299**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,4 +1,13 @@
...
   include SessionsHelper
+
+  private
+    def logged_in_user
+      unless logged_in?
+        store_location
+        flash[:danger] = "Please log in."
+        redirect_to login_url
+      end
+    end
```
Comment (b-cuong) [inline app/controllers/application_controller.rb]: indent. CHẠY RUBOCOP CHƯA?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,4 +1,13 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   include SessionsHelper
+
+  private
+    def logged_in_user
+      unless logged_in?
+        store_location
+        flash[:danger] = "Please log in."
```
Comment (b-cuong) [inline app/controllers/application_controller.rb]: i18n

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,20 +1,56 @@
 class UsersController < ApplicationController
-  def show
-    @user = User.find(params[:id])
-  end
+  before_action :set_user, except:[:index, :new, :create]
```
Comment (b-cuong) [inline app/controllers/users_controller.rb]: thêm space sau `except:`

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,20 +1,56 @@
...
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
```
Comment (b-cuong) [inline app/controllers/users_controller.rb]: `except`

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,20 +1,56 @@
...
+  before_action :admin_user, only: :destroy
+
+  def show; end
 
   def new
     @user = User.new
   end
 
+  def index
+    @users = User.paginate(page: params[:page])
```
Comment (b-cuong) [inline app/controllers/users_controller.rb]: bỏ ()

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,20 +1,56 @@
...
+  end
+
+  def edit; end
+
+  def update
+    if @user.update_attributes(user_params)
```
Comment (b-cuong) [inline app/controllers/users_controller.rb]: bỏ ()

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -1,20 +1,56 @@
...
+    redirect_to users_url
+  end
+
   private
 
-  def user_params
-    params.require(:user).permit(:name, :email, :password, :password_confirmation)
-  end
+    def set_user
+      @user = User.find(params[:id])
```
Comment (b-cuong) [inline app/controllers/users_controller.rb]: ()

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -1,20 +1,56 @@
...
+    end
+
+    def user_params
+      params.require(:user).permit(:name, :email, :password, :password_confirmation)
```
Comment (b-cuong) [inline app/controllers/users_controller.rb]: ()

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -1,20 +1,56 @@
...
+    end
+
+    def correct_user
+      redirect_to(root_url) unless current_user?(@user)
```
Comment (b-cuong) [inline app/controllers/users_controller.rb]: ()

File: app/helpers/users_helper.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+module UsersHelper
+  # Returns the Gravatar for the given user.
```
Comment (b-cuong) [inline app/helpers/users_helper.rb]: bỏ comment thừa

File: app/helpers/users_helper.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+  def gravatar_for(user, options = { size: 80 })
```
Comment (b-cuong) [inline app/helpers/users_helper.rb]: ()

File: app/helpers/users_helper.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    size = options[:size]
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}?s=#{size}"
+    image_tag(gravatar_url, alt: user.name, class: "gravatar")
```
Comment (b-cuong) [inline app/helpers/users_helper.rb]: ()

File: app/views/users/edit.html.erb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+<% provide(:title, t(".edit_user")) %>
+<% provide(:button_text, t(".save")) %>
```
Comment (b-cuong) [inline app/views/users/edit.html.erb]: ()

File: app/views/users/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+<h1><%= t(".update_profile")%></h1>
```
Comment (b-cuong) [inline app/views/users/edit.html.erb]: ()

File: app/views/users/edit.html.erb [hunk 15]
```diff
@@ -0,0 +1,8 @@
...
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= render 'form' %>
```
Comment (b-cuong) [inline app/views/users/edit.html.erb]: ""

File: app/views/users/index.html.erb [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<% provide(:title, t(".all_users")) %>
```
Comment (b-cuong) [inline app/views/users/index.html.erb]: ()

File: app/views/users/index.html.erb [hunk 17]
```diff
@@ -0,0 +1,7 @@
...
+<h1><%= t(".all_users") %></h1>
```
Comment (b-cuong) [inline app/views/users/index.html.erb]: ()

File: app/views/users/new.html.erb [hunk 18]
```diff
@@ -1,4 +1,5 @@
 <% provide(:title, t(".signup")) %>
+<% provide(:button_text, t(".textsubmit")) %>
```
Comment (b-cuong) [inline app/views/users/new.html.erb]: ()

File: app/views/users/new.html.erb [hunk 19]
```diff
@@ -1,4 +1,5 @@
...
 <h1><%= t(".signup") %></h1>
```
Comment (b-cuong) [inline app/views/users/new.html.erb]: ()

File: app/views/users/show.html.erb [hunk 20]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, @user.name) %>
```
Comment (b-cuong) [inline app/views/users/show.html.erb]: ()

File: db/seeds.rb [hunk 21]
```diff
@@ -1,7 +1,14 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
-#
-# Examples:
-#
-#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
-#   Character.create(name: 'Luke', movie: movies.first)
+User.create!(name:  "Example User",
+             email: "example@railstutorial.org",
+             password:              "foobar",
```
Comment (b-cuong) [inline db/seeds.rb]: space

File: db/seeds.rb [hunk 22]
```diff
@@ -1,7 +1,14 @@
...
+             password_confirmation: "foobar")
+
+99.times do |n|
+  name  = Faker::Name.name
+  email = "example-#{n+1}@railstutorial.org"
+  password = "password"
+  User.create!(name:  name,
+               email: email,
+               password:              password,
```
Comment (b-cuong) [inline db/seeds.rb]: space

File: app/views/users/_user.html.erb [hunk 23]
```diff
@@ -0,0 +1,7 @@
+<li>
+  <%= gravatar_for user, size: 50 %>
+  <%= link_to user.name, user %>
+  <% if current_user.admin? && !current_user?(user) %>
+    | <%= link_to "delete", user, method: :delete,data: { confirm: "You sure?" } %>
```
Comment (duongpham910) [inline app/views/users/_user.html.erb]: i18n

File: app/helpers/sessions_helper.rb [hunk 24]
```diff
@@ -25,6 +25,10 @@ def logged_in?
     !current_user.nil?
   end
 
+  def current_user?(user)
```
Comment (duongpham910) [inline app/helpers/sessions_helper.rb]: bỏ ()

**Pull request merged**